packages/amplify-cli-core/src/state-manager/stateManager.ts

```diff
@@ -72,8 +105,155 @@ export class StateManager {
+        const policyWithEnv = this.replaceEnvForCustomPolicies(policy, envName);
+        customPolicies.push(policyWithEnv);
+      }
+      else {
+        printer.warn(`The format of custom IAM policies in the ${categoryName} ${resourceName} is not valid`);
+        return undefined;
+      }
+    }
+    //the policies without env will be carried over and merge into the CFN template
+    return customPolicies;
+  };
+
+  //validate the format of actions and ARNs for custom IAM policies
+  validateRegexCustomPolicy = (customPolicy: CustomIAMPolicy, resourceName: string) : void => {
+    const resources = customPolicy.Resource;
+    const actions = customPolicy.Action;
+    let resourceRegex = new RegExp('(arn:(aws[a-zA-Z0-9-]*):([a-zA-Z0-9\\-])+:([a-z]{2}(-gov)?-[a-z]+-\\d{1})?:(\\d{12})?:(.*))*');
+    let actionRegex = new RegExp('([a-z0-9])*:([a-z|A-Z|0-9|*]+)*');
+    let wrongResourcesRegex = [];
+    let wrongActionsRegex = [];
+    let errorMessage = "";
+    const printer: Printer = new AmplifyPrinter()
+
+    for (const resource of resources) {
+      if (!resourceRegex.test(resource)) {
+        wrongResourcesRegex.push(resource);
+      }
+    }
+
+    if(resources.includes('*')) {
+      printer.warn(`Warning: You've specified "*" as a custom IAM policy for your ${resourceName}. 
+      This will give your ${resourceName} access to ALL resources in the AWS Account.`)
+    }
+
+    for (const action of actions) {
+      if (!actionRegex.test(action)) {
+        wrongActionsRegex.push(action);
+      }
+    }
+
+    if (wrongResourcesRegex.length > 0) {
+      errorMessage += `\nInvalid ARN format for custom IAM policies in ${resourceName}:\n${wrongResourcesRegex.toString()}\n`;
+    }
+    if (wrongActionsRegex.length > 0) {
+      errorMessage += `\nInvalid actions format for custom IAM policies in ${resourceName}:\n${wrongActionsRegex.toString()}\n`;
+    }
+
+    if (errorMessage.length > 0) {
+      printer.error(errorMessage);
+    }
+  }
+
+
+  //replace or add env parameter in the front of the resource customers enter to the current env
+  replaceEnvForCustomPolicies = (policy: CustomIAMPolicy, currentEnv: string): CustomIAMPolicy => {
+    let resourceWithEnv = [];
+    let action = policy.Action;
+    let effect = policy.Effect;
+    let customIAMpolicies: CustomIAMPolicy = {
+      Action: action,
+      Effect: effect,
+      Resource: []
+    }
+    for (let resource of policy.Resource){
+      resourceWithEnv.push(resource.replace( '{env}', currentEnv));
+    }
+    customIAMpolicies.Resource = resourceWithEnv;
+    return customIAMpolicies;
+  }
+
+  addCustomPoliciesFile = (categoryName: string, resourceName: string): void => {
+    const customPoliciesPath = pathManager.getCustomPoliciesPath(categoryName, resourceName);
+    const defaultCustomPolicies = {
+        policies: [
+          {
+            Effect: 'Allow',
+            Action: [],
+            Resource: []
+          }
+        ]
+    }
```

Comment:
We cannot use cloudform type for the default value of custom policies file. Because in our design doc, it requires us to have resource and action in the template. But using cloudform type iam.policyStatement, if there is no value in the action or resource, then the action and resource won’t be generated in the template. And we cannot add null value or empty string in the action or resource, for there will be error.